### ISSUE TITLE
Add support non-default (hive) catalog for Hive Metastore 3.x or newer

### ISSUE THREAD

**Issue by dmitryfill** [trinodb/trino#10287]
Current `hive` (former `hive-hadoop2`) connector and `iceberg` especially need to support ability to set different catalog from default (`hive` catalog) in the Hive Metastore 3.x or newer.

**Comment by findinpath** [comment]
When using `CREATE SCHEMA`
you could create different schemas within the hive/iceberg connector.
https://trino.io/docs/current/connector/hive.html#examples

If this is not what you are looking for, please add more details to the issue description.

**Comment by dmitryfill** [comment]
Default catalog for HiveMetastore client is `hive`. Seems like starting from Spark 3.x there is a possibility to provide different default catalog, for example (https://spark.apache.org/docs/latest/configuration.html)

```
spark.sql.catalog.spark_catalog=xxx
spark.sql.defaultCatalog=xxx
```

with adoption of this across Data Lakes and Iceberg (described here https://iceberg.apache.org/#spark-configuration/#catalogs) Trino needs to support via config ability to setup connectors for custom catalogs (non default `hive` catalogs)

**Comment by findinpath** [comment]
Can you please script in more detail your usage scenario?

Currently Trino allows its users to use multiple connectors (hive & iceberg) with the same hive metastore.

If you look at the catalog configuration for:
- [hive](https://github.com/trinodb/trino/blob/master/testing/trino-server-dev/etc/catalog/hive.properties)
- [iceberg](https://github.com/trinodb/trino/blob/master/testing/trino-server-dev/etc/catalog/iceberg.properties)

in the `trino-server-dev` you can notice that they use the same HMS

```
hive.metastore.uri=thrift://localhost:9083
```


If you use the command (in Trino base directory):

```
testing/bin/ptl env up --environment 'singlenode'  --config 'config-default'
```

and then try to do in SQL:

```
create table hive.default.hivetable (id bigint);

create table iceberg.default.icebergtable (id bigint);
```

This will work as expected.

**Comment by hashhar** [comment]
@findinpath Hive Metastore since 3.x has the ability to have "catalogs" as well which is what @dmitryfill is pointing to if I understand correctly.

Today if you have a Hive Metastore with two Hive catalogs - `hive` and `mynewhive` then Trino only looks at the default catalog `hive`.

IMO the way to support this would require having multiple hive catalogs in Trino - 1 for each catalog in HMS.

**Comment by findepi** [comment]
> IMO the way to support this would require having multiple hive catalogs in Trino - 1 for each catalog in HMS.

Which boils down to adding a configuration toggle, passing it appropriately in HMS calls, with proper tests.
Idk however, how to handle this nicely, and remain compatible with HMS < 3. To be found out.

**Comment by dmitryfill** [comment]
Yes, @hashhar is correct. Hive Metastore 3.x supports custom catalogs, pretty much like a different namespaces for databases and tables.

As of scenario, lets assume I have 2 different environments `staging` & `production`

In Trino 2 hive connectors added:

1. Staging

```properties
hive.metastore.uri=thrift://192.0.2.3:9083
hive.metastore.catalog.default=hive_staging
```

2. Production
```properties
hive.metastore.uri=thrift://192.0.2.3:9083
hive.metastore.catalog.default=hive_production
```

And there is a SQL hardcoded in the code: 

```sql
SELECT col1, col2, col3 FROM database_foo.table_bar
WHERE col1 BETWEEN 7 ANF 25
```

catalog now could be specified through configuration, both in Spark or in Trino, (in Trino use case just a hive connector with specified catalog in Hive Metastore to connect as example above), for example;

```java
  String hiveCatalog = System.getenv("ENV_HIVE_CATALOG");
  String url = String.format("jdbc:trino://example.net:8443/%s?user=test&password=secret&SSL=true", hiveCatalog);
  Connection connection = DriverManager.getConnection(url);

  // Execute SQL above
```

Same code could be run now in different environments, and this gives a decent code portability.

---

More references in `HiveMetaStoreClient`

- `getDatabases` for specific catalog
https://github.com/apache/hive/blob/3c9ea5d8d6ecc0d3749e22b28bfb7690dd3f1be1/standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java#L1958-L1963

**Comment by dmitryfill** [comment]
Any updates on this? Is that something feasible?

**Comment by osscm** [comment]
Wondering, one option could be to extend [ThriftHiveMetastoreClient](https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastoreClient.java) as `ThriftHive3MetastoreClient` which can accept the `hive catalog`, this can allow to use different `ThriftMetastoreClient`. Though not sure how neatly different clients can be bound from the `Factory/Module/PluginFactory`. Or if not then can have Hive3 based connectors by just overriding `Factory/Module/PluginFactory`.

**Comment by alexandre-cardoso** [comment]
Any thoughts on this? I have configured a Trino 402 cluster with 2 different catalogs (hive and hudi), both pointing to the same HMS. Given that, when I prompt to show schemas from each catalog, Trino outputs the same schemas because all schemas were created in the hive default catalog in HMS. Is there a way to schema HMS custom catalog per Trino catalog?

**Comment by osscm** [comment]
@alexandre-cardoso I will take up this issue as may be making it 1:1 mapped with Trino catalog is a good idea. 
making sure that it will work with hive 1.x as well.

Side note, I had tested with the changes and it works in the local.

**Comment by dmitryfill** [comment]
@findepi , @hashhar , @osscm - Any updates on this one?

**Comment by osscm** [comment]
@findepi do you have any thoughts on this based on the above discussion.

I was planning to keep it simple as 1:1 mapping between Trino catalog to Hive catalog, and based on the toggle we can use the right  `ThriftHiveMetastoreClient`. Thrift API already supports `catalog` for the operations.

**Comment by cedwards-az** [comment]
Hi, I'm looking for some movement around this?
If the answer is a single HMS per catalogue, so be it, but an positioning statement would help.
I can accept both answers, but looking for best practice guidance

**Comment by rokyaria** [comment]
Is this issue still not resolved?

**Comment by hashhar** [comment]
cc: @electrum since this change would impact the Hadoop decoupling probably.

**Comment by electrum** [comment]
This won’t impact Hadoop decoupling directly, but will be significant work and likely can’t be done in backwards compatible way without even more work.

**Comment by Tyron2308** [comment]
hello, any update on this ?

**Comment by maxpoulain** [comment]
Hello !

Is there any news on that topic ? 
In my case, I would like to have 3 catalogs (raw / silver / gold) with only one HMS ! In particular to manage user rights more easily and also to avoid to maintain 3 HMS..

**Comment by samssh** [comment]
Hello everyone. Has there been any progress on this matter? Is there anything I can do to help resolve this issue faster?

**Comment by hashhar** [comment]
Add a new config property for Hive connector to specify the HMS catalog for a given Trino catalog.
Use the config when available to pass it in all the Thrift calls where the catalog parameter is needed.
Add tests against Hive 4 (requires building a docker based environment like in https://github.com/trinodb/docker-images/tree/master/testing/hive3.1-hive) to verify this works.

**Comment by samssh** [comment]
What about importing `hive-jdbc` in the `trinodb/hive` repository and using `IMetaStoreClient`? The implementation of this interface internally handles the catalog name via the "metastore.catalog.default" configuration property. It is also compatible with Hive 2.x and 3.x, I suppose.

And one more thing, why do we need a test for Hive 4? The catalog name was introduced in Hive 3, and supporting Hive 4 is beyond the scope of this issue, right?

**Comment by anusudarsan** [comment]
> And one more thing, why do we need a test for Hive 4? The catalog name was introduced in Hive 3, and supporting Hive 4 is beyond the scope of this issue, right?

Thats right. The hive "catalog" support can be done with Hive 3. Hive 4 upgrade is orthogonal and is out of scope for this issue

**Comment by osscm** [comment]
ref: https://github.com/trinodb/trino/issues/10287#issuecomment-1426688247

@hashhar we have implemented it internally, but initially did not see much interest, so didn't put effort to push it upstream.
We made it backward compatible by extending `ThriftHiveMetastoreClient` for the custom catalog (It is optional, but this way we kept it clean and did not bother about the regression).
We have also made it config driven, as how you have suggested.
We are using this feature production for a few years.

It seems, this will help community, so will request your guidance to push it to upstream.

@samssh please see, if you are on slack, we can discuss it there, thanks?

**Comment by hashhar** [comment]
Since multiple people are working on this: @anusudarsan @osscm @samssh , it would be to coordinate amongst yourselves to make sure nobody is doing repeated work and you can assist with code reviews or suggestions.

Please also get in touch with @electrum / @dain to make sure the change aligns with how he expects the metastore interface to evolve.

**Comment by samssh** [comment]
I've joined the Slack workspace and we'll be discussing there, but I'm not sure where to start. @osscm, could you guide me, please?

**Comment by dain** [comment]
Reading the code linked above, it appears the HMS has decided to use the following [hack to encode catalog name into existing database name parameter](https://github.com/apache/hive/blob/master/standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreUtils.java#L1018-L1035):
```
@<catalogName>#<databaseName>
```
If the database name is empty a `!` is used instead

**Comment by osscm** [comment]
> I've joined the Slack workspace and we'll be discussing there, but I'm not sure where to start. @osscm, could you guide me, please?

hi @samssh please see.
pinged you over slack (may be you are offline), and also added you to the discussion, please check.
based on the discussion with @dain and @anusudarsan and their suggestion, have started this PR: https://github.com/trinodb/trino/pull/21502 

Quick gist:

They wanted me to have an initial PR, for them to understand what changes we did internally.
and based on that Dain's and Anu suggested, we can do similar changes in the existing client and extend the PR.
